plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenShiftProjectCache.java

```diff
@@ -50,79 +52,119 @@ public OpenShiftProjectCache() {
 	}
 	
 	@Override
-	public synchronized Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
+	public Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
 		String key = getCacheKey(conn);
-		if(!cache.containsKey(key)) {
+		Set<IProjectAdapter> adapters = null;
+		if (!contains(key)) {
+			adapters = createEntry(key);
 			Collection<IProject> projects = conn.<IProject>getResources(ResourceKind.PROJECT);
-			Set<IProjectAdapter> adapters = new HashSet<>();
 			for (IProject project : projects) {
-				newProjectAdapter(adapters, conn, project);
+				IProjectAdapter adapter = addNewProjectAdapter(adapters, conn, project);
+				notifyAdd(adapter);
 			}
-			cache.put(key, adapters);
-			notifyAdd(adapters);
+		} else {
+			adapters = getEntry(key);
 		}
-		return new ArrayList<>(cache.get(key));
+
+		return new ArrayList<>(adapters);
```

Comment:
I dont think so. Modifying the adapters is only done on an new (cache-) entry. Creating the entry is locking, thus 2 thread wont be able to both see the entry as being new.